File: Python Fundamentals/9. Regular Expression/furniture.py
import re

pattern = r">>(?P<name>[A-Za-z]+)<<(?P<price>\d+(\.?\d+)?)\!(?P<quantitiy>\d+)"
bought_furniture = []
total_cost = 0

txt = input()

while not txt == "Purchase":
    action = re.match(pattern, txt, re.IGNORECASE)
    if action:
        el_dict = action.groupdict()
        bought_furniture.append(el_dict["name"])
        total_cost += float(el_dict["price"]) * int(el_dict['quantitiy'])
    txt = input()

print("Bought furniture:")
if bought_furniture:
    print("\n".join(bought_furniture))
print(f"Total money spend: {total_cost:.2f}")

File: JavaScript Fundamentals/JavaScript Fundamentals Regex/furniture.js
function solve(input) {
    let pattern = />>\b(?<name>[A-Za-z]+)<<(?<price>\d+|\d+\.\d+)\!(?<quantity>\d+)/g;
    let bought = [];
    let totalMoney = 0;
    
    for (let el of input) {
        if(el.match(pattern)) {
            let match = pattern.exec(el);
            let {name, price, quantity} = match.groups;
            bought.push(name);
            totalMoney += Number(price) * Number(quantity);
        }
    }

    console.log('Bought furniture:');
    if (bought.length > 0) {
        console.log(bought.join('\n'));
    }
    console.log(`Total money spend: ${totalMoney.toFixed(2)}`);
}

// solve(['>>Laptop<<312.2323!3',
// '>>TV<<300.21314!5',
// '>Invalid<<!5',
// '>>TV<<300.21314!20',
// '>>Invalid<!5',
// '>>TV<<30.21314!5',
// '>>Invalid<<!!5',
// 'Purchase']);

// solve(['>>Sofa<<312.23!3',
// '>>TV<<300!5',
// '>Invalid<<!5',
// 'Purchase']);

solve(['>Invalid<<!4',
'>Invalid<<!2',
'>Invalid<<!5',
'Purchase']);